### ISSUE TITLE
False positive when destructuring: Object is possibly null

### ISSUE THREAD

**Issue by dlindenkreuz** [Microsoft/TypeScript#16175]
**TypeScript Version:**  2.3.4

**Code**

```ts
type Foo = {
    bar: {
        baz: number | null
    } | null
} | null

function works(foo: Foo): number {
    if (!foo || !foo.bar || !foo.bar.baz) return 0

    const baz = foo.bar.baz
    return baz + 123
}

// should work but does not
function shouldWork(foo: Foo): number {
    if (!foo || !foo.bar || !foo.bar.baz) return 0

    const { baz } = foo.bar
    // following line throws this:
    // → Type '{ baz: number | null; } | null' has no property 'baz' and no string index signature
    // const { bar: { baz } } = foo
    return baz + 123 // baz: Object is possibly 'null'
}
```

**Expected behavior:**
It should not matter whether I extract a property using destructuring or access it through the regular dot notation. Additionally, the commented line with nested destructuring should not give me this error: `Type '{ baz: number | null; } | null' has no property 'baz' and no string index signature`.

**Actual behavior:**
The type checking errors described above are thrown when `"strict": true` is set in tsconfig.json in v2.3.4. The error does not occur in v2.3.2.

**Comment by mhegazy** [comment]
I am unable to reproduce this issue.

```sh
c:\test\16175> type a.ts
type Foo = {
    bar: {
        baz: number | null
    } | null;
} | null;

function works(foo: Foo): number {
    if (!foo || !foo.bar || !foo.bar.baz) return 0

    const baz = foo.bar.baz
    return baz + 123
}

// should work but does not
function shouldWork(foo: Foo): number {
    if (!foo || !foo.bar || !foo.bar.baz) return 0

    const { baz } = foo.bar
    // following line throws this:
    // ΓåÆ Type '{ baz: number | null; } | null' has no property 'baz' and no string index signature
    // const { bar: { baz } } = foo
    return baz + 123 // baz: Object is possibly 'null'
}

c:\test\16175>tsc --v
Version 2.5.0-dev.20170816

c:\test\16175>tsc a.ts --strict

c:\test\16175>echo %errorlevel%
0
```

**Comment by dlindenkreuz** [comment]
Thanks for chiming in.

The line that throws the error was unfortunately commented out, sorry for providing a non-throwing snippet. Here is the version that throws using tsc v2.3.4 (v2.4.x as well) and `--strict`:

```
function shouldWork(foo: Foo): number {
    if (!foo || !foo.bar || !foo.bar.baz) return 0

    // const { baz } = foo.bar
    // following line throws this:
    // → Type '{ baz: number | null; } | null' has no property 'baz' and no string index signature
    const { bar: { baz } } = foo
    return baz + 123 // baz: Object is possibly 'null'
}
```

However, the error seems to be gone using Typescript >= v2.5 🎉
